File: top_interview_questions_easy/LongestCommonPrefix.py
class Solution(object):
    def longestCommonPrefix(self, strs):
        if not strs:
            return ""
        
        # zip() : 여러개의 순회 가능한 객체를 인자로 받고 각 객체가 담고있는 원소를 
        # 튜플 형태로 차례로 접근할 수 있는 반복자 반환
        for i, letter_group in enumerate(zip(*strs)):
          # *변수: 해당 매개변수를 가변적인 갯수를 가진 위치 인수로 정의하겠다는 것
          참고: https://hcnoh.github.io/2019-01-27-python-arguments-asterisk
            if len(set(letter_group)) >1:
                return strs[0][:1]
            else:
                return min(strs)
        

File: top_interview_questions_easy/longest_common_prefix.cpp
class Solution {
public:
    string longestCommonPrefix(vector<string>& strs) {
      if(strs.size() == 0 ) return "";
      // strs을 사전식으로 정렬
      sort(strs.begin(), strs.end());
      // 그 중 첫번째와 마지막 문자를 가져옴
      string frontString = strs[0], lastString = strs[strs.size() - 1], result = "";
      // 각 string 배열의 길이를 가져옴
      int frontSize = frontString.size(), lastSize = lastString.size();
      
      // 0부터 시작해서 frontSize와 lastSize보다 i가 작을 동안
      // 앞부터 차례대로 가져오는 frontString과, 뒤에서 차례로 가져오는 lastString을 비교하여 
      // 같을 경우 result에 front[i] 해당 문자를 삽입
      for(int i = 0; i < frontSize && i < lastSize && frontString[i] == lastString[i]; i++) {
        result += frontString[i];
      }
      return result;
    }
};